### PR TITLE
K8SPSMDB-327: Add mongos and cfg server sans to certificates

### PR DESCRIPTION
**Pull request by qJkee**
[K8SPSMDB-327](https://jira.percona.com/browse/K8SPSMDB-327)

After change, we will have the following certificate DNS names
```
  Dns Names:
    localhost
    my-cluster-name-rs0
    my-cluster-name-rs0.run-k8spsmdb-327-9430
    my-cluster-name-rs0.run-k8spsmdb-327-9430.svc.cluster.local
    *.my-cluster-name-rs0
    *.my-cluster-name-rs0.run-k8spsmdb-327-9430
    *.my-cluster-name-rs0.run-k8spsmdb-327-9430.svc.cluster.local
    my-cluster-name-mongos
    my-cluster-name-mongos.run-k8spsmdb-327-9430
    my-cluster-name-mongos.run-k8spsmdb-327-9430.svc.cluster.local
    *.my-cluster-name-mongos
    *.my-cluster-name-mongos.run-k8spsmdb-327-9430
    *.my-cluster-name-mongos.run-k8spsmdb-327-9430.svc.cluster.local
    my-cluster-name-cfg
    my-cluster-name-cfg.run-k8spsmdb-327-9430
    my-cluster-name-cfg.run-k8spsmdb-327-9430.svc.cluster.local
    *.my-cluster-name-cfg
    *.my-cluster-name-cfg.run-k8spsmdb-327-9430
    *.my-cluster-name-cfg.run-k8spsmdb-327-9430.svc.cluster.local
```

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| arbiter | passed |
| security-context | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-485-9947f59b

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| storage | passed |
| monitoring | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by delgod — APPROVED** [review]

File: pkg/controller/perconaservermongodb/ssl.go [hunk 1]
```diff
@@ -209,6 +211,23 @@ func (r *ReconcilePerconaServerMongoDB) createSSLManualy(cr *api.PerconaServerMo
...
 
+func getShardingSans(clusterName, namespace, clusterSuffix string) []string {
+	return []string{
+		clusterName + "-mongos",
+		clusterName + "-mongos" + "." + namespace,
+		clusterName + "-mongos" + "." + namespace + "." + clusterSuffix,
+		"*." + clusterName + "-mongos",
+		"*." + clusterName + "-mongos" + "." + namespace,
+		"*." + clusterName + "-mongos" + "." + namespace + "." + clusterSuffix,
+		clusterName + "-cfg",
```
Comment (blez) [inline pkg/controller/perconaservermongodb/ssl.go:222]: We have `api.ConfigReplSetName` const.
Maybe should be added same for mongos

**Review by blez — CHANGES_REQUESTED** [review]

**Review by tplavcic — CHANGES_REQUESTED** [review]
As discussed we should probably re-generate our test/example ssl secrets to support this also: https://github.com/percona/percona-server-mongodb-operator/blob/master/deploy/ssl-secrets.yaml

File: pkg/controller/perconaservermongodb/ssl.go (hunk 1 above)
Comment (qJkee) [reply pkg/controller/perconaservermongodb/ssl.go]: yea, good point
fixed in 
af56d17

**Review by delgod — DISMISSED** [review]

**Review by delgod — APPROVED** [review]

File: pkg/controller/perconaservermongodb/ssl.go [hunk 2]
```diff
@@ -209,6 +211,23 @@ func (r *ReconcilePerconaServerMongoDB) createSSLManualy(cr *api.PerconaServerMo
 	return nil
 }
 
+func getShardingSans(cr *api.PerconaServerMongoDB) []string {
```
Comment (blez) [inline pkg/controller/perconaservermongodb/ssl.go:214]: Not sure, but maybe we should check if sharding is enabled.

Comment (qJkee) [reply pkg/controller/perconaservermongodb/ssl.go:214]: as we discussed, we always add this since sharding can be enabled later

Comment (delgod) [reply pkg/controller/perconaservermongodb/ssl.go:214]: yes, we add them always to not REgenerate when somebody enables sharding

Comment (blez) [reply pkg/controller/perconaservermongodb/ssl.go:214]: then instead of `cr.Spec.Sharding.ConfigsvrReplSet.Name` use `api.ConfigReplSetName`

**Review by tplavcic — DISMISSED** [review]

File: pkg/controller/perconaservermongodb/ssl.go (hunk 2 above)
Comment (delgod) [reply pkg/controller/perconaservermongodb/ssl.go:214]: fixed in https://github.com/percona/percona-server-mongodb-operator/pull/485/commits/4e5cc3047aa62b895b7e1a41b26fc02bfbe2a526

**Review by blez — APPROVED** [review]

**Comment by delgod** [comment]
> As discussed we should probably re-generate our test/example ssl secrets to support this also: https://github.com/percona/percona-server-mongodb-operator/blob/master/deploy/ssl-secrets.yaml

Hi @tplavcic 
we created certificates for `psmdb` namespace
```
localhost, DNS
my-cluster-name-rs0, DNS
my-cluster-name-rs0.psmdb, DNS
my-cluster-name-rs0.psmdb.svc.cluster.local, DNS
*.my-cluster-name-rs0, DNS
*.my-cluster-name-rs0.psmdb, DNS
*.my-cluster-name-rs0.psmdb.svc.cluster.local, DNS
my-cluster-name-mongos, DNS
my-cluster-name-mongos.psmdb, DNS
my-cluster-name-mongos.psmdb.svc.cluster.local, DNS
*.my-cluster-name-mongos, DNS
*.my-cluster-name-mongos.psmdb, DNS
*.my-cluster-name-mongos.psmdb.svc.cluster.local, DNS
my-cluster-name-cfg, DNS
my-cluster-name-cfg.psmdb, DNS
my-cluster-name-cfg.psmdb.svc.cluster.local, DNS
*.my-cluster-name-cfg, DNS
*.my-cluster-name-cfg.psmdb, DNS
*.my-cluster-name-cfg.psmdb.svc.cluster.local
```
in https://github.com/percona/percona-server-mongodb-operator/pull/485/commits/9db0184d274684e6264772694c07b83f82b9fc21

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-485-4e5cc304

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| arbiter | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | failed |
| init-deploy | passed |
| upgrade | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | failed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |
| monitoring-2-0 | failed |
| self-healing | passed |
| operator-self-healing | passed |
| arbiter | failed |
| service-per-pod | failed |
| liveness | failed |
| smart-update | failed |
| version-service | failed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | failed |
| one-pod | passed |
| upgrade | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | failed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |
| monitoring-2-0 | failed |
| self-healing | passed |
| operator-self-healing | passed |
| arbiter | failed |
| service-per-pod | failed |
| liveness | failed |

**Pull request merged**